hardware_interface/src/resource_manager.cpp

```diff
@@ -51,116 +81,416 @@ class ResourceStorage
+      // TODO(destogl): is it better to check here if previous state was unconfigured instead of
+      // checking if each state already exists? Or we should somehow know that transition has
+      // happened and only then trigger this part of the code?
+      // On the other side this part of the code should never be executed in real-time critical
+      // thread, so it could be also OK as it is...
+      // @bmagyar what do you think?
+      for (const auto & interface : hardware_info_map_[hardware.get_name()].state_interfaces)
+      {
+        // add all state interfaces to available list
+        auto found_it = std::find(
+          available_state_interfaces_.begin(), available_state_interfaces_.end(), interface);
+
+        if (found_it == available_state_interfaces_.end())
+        {
+          available_state_interfaces_.emplace_back(interface);
+          // TODO(destogl): change output to debug
+          RCUTILS_LOG_INFO_NAMED(
+            "resource_manager", "(hardware '%s'): '%s' state interface added into available list",
+            hardware.get_name().c_str(), interface.c_str());
+        }
+        else
+        {
+          // TODO(destogl): do here error management if interfaces are only partially added into
+          // "available" list - this should never be the case!
+          RCUTILS_LOG_WARN_NAMED(
+            "resource_manager",
+            "(hardware '%s'): '%s' state interface already in available list."
+            " This can happen due to multiple calls to 'configure'",
+            hardware.get_name().c_str(), interface.c_str());
+        }
+      }
+
+      // add command interfaces to available list
+      for (const auto & interface : hardware_info_map_[hardware.get_name()].command_interfaces)
+      {
+        // TODO(destogl): check if interface should be available on configure
+
+        auto found_it = std::find(
+          available_command_interfaces_.begin(), available_command_interfaces_.end(), interface);
+
+        if (found_it == available_command_interfaces_.end())
+        {
+          available_command_interfaces_.emplace_back(interface);
+          // TODO(destogl): change output to debug
+          RCUTILS_LOG_INFO_NAMED(
+            "resource_manager", "(hardware '%s'): '%s' command interface added into available list",
+            hardware.get_name().c_str(), interface.c_str());
+        }
+        else
+        {
+          // TODO(destogl): do here error management if interfaces are only partially added into
+          // "available" list - this should never be the case!
+          RCUTILS_LOG_WARN_NAMED(
+            "resource_manager",
+            "(hardware '%s'): '%s' command interface already in available list."
+            " This can happen due to multiple calls to 'configure'",
+            hardware.get_name().c_str(), interface.c_str());
+        }
+      }
+    }
+    return result;
+  }
+
+  template <class HardwareT>
+  bool cleanup_hardware(HardwareT & hardware)
```

Comment:
do you have ideas on how to do this with less code? the vast part of this code seems to be following the same pattern as above for `configure`.

Reply:
TBH no. The code is very similar but with some subtle differences. Putting this into a separate (lambda) method would make this method complex or arguments would be hard to understand. For clarity, I would keep this longer code.

Reply:
ok